opnode/l1/downloader.go

```diff
@@ -0,0 +1,220 @@
+package l1
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	lru "github.com/hashicorp/golang-lru"
+
+	"github.com/ethereum-optimism/optimistic-specs/opnode/eth"
+
+	"github.com/ethereum/go-ethereum"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/event"
+)
+
+type blockAndReceipts struct {
+	// Track if we finished (with error or not). May increment >1 when other sub-tasks fail.
+	// First field, aligned atomic changes.
+	finished uint32
+
+	// Count receipts to track status
+	// First field of struct for memory aligned atomic access
+	DownloadedReceipts uint32
+
+	Block *types.Block
+	// allocated in advance, one for each transaction, nil until downloaded
+	Receipts []*types.Receipt
+
+	// stop fetching if this context is dead
+	ctx context.Context
+
+	// for other duplicate requests to get the result
+	feed event.Feed
+}
+
+// wrappedErr wraps an error, since event.Feed cannot handle nil errors otherwise (reflection on nil)
+type wrappedErr struct {
+	error
+}
+
+func (bl *blockAndReceipts) Finish(err wrappedErr) {
+	if atomic.AddUint32(&bl.finished, 1) == 1 {
+		bl.feed.Send(err)
+	}
+}
+
+type receiptTask struct {
+	BlockHash common.Hash
+	TxHash    common.Hash
+	TxIndex   uint64
+	// Count the attempts we made to fetch this receipt. Block as a whole fails if we tried to many times.
+	Retry uint64
+	// Avoid concurrent Downloader cache access and pruning edge cases with receipts
+	// Keep a pointer to insert the receipt at
+	Dest *blockAndReceipts
+}
+
+type Downloader interface {
+	Fetch(ctx context.Context, id eth.BlockID) (*types.Block, []*types.Receipt, error)
+	AddReceiptWorkers(n int) int
+}
+
+type DownloadSource interface {
+	eth.BlockByHashSource
+	eth.ReceiptSource
+}
+
+type downloader struct {
+	// cache of ongoing/completed block tasks: block hash -> block
+	cacheEvict *lru.Cache
+	cacheLock  sync.Mutex
+
+	receiptTasks       chan *receiptTask
+	receiptWorkers     []ethereum.Subscription
+	receiptWorkersLock sync.Mutex
+
+	chr DownloadSource
+}
+
+var downloadEvictedErr = errors.New("evicted")
+
+func NewDownloader(chr DownloadSource) Downloader {
+	dl := &downloader{
+		receiptTasks: make(chan *receiptTask, 100),
+		chr:          chr,
+	}
+	evict := func(k, v interface{}) {
+		// stop downloading things if they were evicted (already finished items are unaffected)
+		v.(*blockAndReceipts).Finish(wrappedErr{downloadEvictedErr})
```

Comment:
> If new request tend to supercede/replace old requests, then the current behaviour is fine.

Generally they should, since we sync is a linear process. With multiple engines it may be more of a risk, but the cache size is 500 block entries, which should be more than plenty.

Reply:
Related edge-case here: on eviction we finish the task, but we should cancel the task context as well, so we free up resources

Reply:
Mmh, I guess the annoying bit is that the conjugation of concerns: the cache is both a cache but also the mechanism that prevents us from having too many requests running at the same time. If we had some form of task queue, then we could reject new requests while the queue is full.

Reply:
We actually do have that: the workers consume from a task queue (`receiptTasks` channel). When the workers start applying back-pressure (full channel, sending a new task will block when consumers are slower than producer and capacity is full) the main `Fetch` calls get blocked. Eviction of a running task should only happen if there's one super slow worker and other parallel fast workers that keep the download cache going.